### PR TITLE
fix(nvim_buf_lines_event): don't send change twice after line truncated.

### LINKED ISSUE #11591: nvim_buf_lines_event: wrong edits received
- `nvim --version`: 0.4.3 / v0.5.0-253-g138480939
- `vim -u DEFAULTS` (version: ) behaves differently?
- Operating system/version: ArchLinux latest in WSL
- Terminal name/version: no applicable
- `$TERM`: no applicable

initial text:
```
abc
{
def
}
```
type: `yyjjvi{p`

receives following edits from nvim_buf_lines_event (`[buf, tick, firstLine, lastLine, lineData, more]`)
```json
[
 [1, 1, 2, 3, [""], 0],
 [1, 2, 2, 3, ["}"], 0], // <- here is the problem
 [1, 3, 3, 4, [], 0],
 [1, 4, 3, 3, ["abc", "}"], 0]
]
```

The result after applying these edits is
```
abc
{
}
abc
}
```
rather than:
```
abc
{

abc
}
```

I'm not sure if i don't miss anything here and it's a bug. Pressing `u` followed by `r` returns expected changes. Terminal nvim behaves as expected

### PR DESCRIPTION
**Pull request by glepnir**
Fix #11591

### CODE REVIEW

**Comment by glepnir** [comment]
See my comment at #11591. `truncate_line` has sent a notification once. `do_join` merges two lines, that is, merges '}' into the third line. At this time the notification was sent again. Actually this is not an error. Because it does change. However, subsequent operations are still being resumed. So you may need to add a condition here to avoid sending notifications or check whether the line is truncated? No shortage of imbeciles.

File: src/nvim/ops.c [hunk 1]
```diff
@@ -3970,7 +3972,9 @@ int do_join(size_t count, int insert_space, int save_undo, int use_formatoptions
   int remove_comments = (use_formatoptions == true)
                         && has_format_option(FO_REMOVE_COMS);
   bool prev_was_comment = false;
-  assert(count >= 1);
+  // don't send change if line is truncated before
+  bool can_send_change = ((curbuf->b_ml.ml_flags & ML_LINE_DIRTY) != 0)
```
Comment (glepnir) [inline src/nvim/ops.c:3976]: Not sure if the conditions here are sufficient or correct. Maybe add a pending operation check?

**Comment by glepnir** [comment]
not sure why `buffer_updates` failed. it pass in my local.

**Comment by glepnir** [comment]
I just checked if the line was truncated before again. Not sure if there are more cases that are broken or need to be considered. do you have any ideas @zeertzjq

**Comment by bfredl** [comment]
I think the issue is rather in `do_put()` . Because it also affects on_bytes . I'm working on a patch for the case it tries to "split" an empty line, which gets the positions wrong.

**Comment by glepnir** [comment]
I'm also not sure if the second send pointed out by the issue author is a bug. Because in dojoin it is merged. In fact it was changed so sending the changes seems to be OK. I'm also not sure if this is normal. Because I don't know what the correct send should be. Glad you're addressing it. close

**Pull request closed without merging**